ref_arch/multi_tenant.rst

```diff
@@ -0,0 +1,590 @@
+
+For example, the advertisements in this application could use a text caption. We can add a column to the table by issuing the standard SQL on the coordinator:
+
+::
+
+  ALTER TABLE ads
+    ADD COLUMN caption text;
+
+This updates all the workers as well. Once this command finishes, the Citus cluster will accept queries that read or write data in the new :code:`caption` column.
+
+For a fuller explanation of which DDL commands propagate through the cluster, see :ref:`ddl_prop_support`.
+
+Per-Tenant Customization
+========================
+
+Given that all tenants share a common schema and hardware infrastructure, how can we accommodate tenants which want to store information not needed by others? For example, one of the tenant applications using our advertising database may want to store tracking cookie information with clicks, whereas another tenant may care about browser agents. Traditionally databases using a shared schema approach for multi-tenancy have resorted to creating a fixed number of pre-allocated "custom" columns, or having external "extension tables." However PostgreSQL provides a much easier way with its unstructured column types, notably `JSONB <https://www.postgresql.org/docs/current/static/datatype-json.html>`_.
+
+Notice that our schema already has a JSONB field in :code:`clicks` called :code:`user_data`. Each tenant can use it for flexible storage.
+
+Supposing company 5 uses the field to track the user agent ad clicks, they can later query to find which browsers click most often:
+
+.. code-block:: postgresql
+
+  SELECT
+    user_data->>'browser' AS browser,
+    count(*) AS count
+  FROM clicks
+  WHERE company_id = 5
+  GROUP BY user_data->>'browser'
+  ORDER BY count DESC
+  LIMIT 10;
+
+The database administrator can even create a `partial index <https://www.postgresql.org/docs/current/static/indexes-partial.html>`_ to improve speed for an individual tenant's query patterns. Here is one to improve company 5's filters for browser information:
+
+.. code-block:: postgresql
+
+  CREATE INDEX click_user_data_browser
+  ON clicks (user_data->>'browser')
+  WHERE company_id = 5;
+
+Additionally, PostgreSQL supports GIN indices on JSONB. Creating a GIN index on a JSONB column will create an index on every key and value within that JSON document. This speeds up a number of `JSONB operators <https://www.postgresql.org/docs/current/static/functions-json.html#FUNCTIONS-JSONB-OP-TABLE>`_ such as :code:`?`, :code:`?|`, and :code:`?&`.
+
+.. code-block:: postgresql
+
+  CREATE INDEX click_user_data
+  ON clicks GIN (user_data);
+
+  -- this allows queries like, "which clicks have
```

Comment:
Also, could we say this speeds up queries like "which clicks .... ". Because the query should be allowed irrespective of whether there is an index or not?